City-of-Helsinki/hel-fi-drupal-grants
# feat: AU-2205, AU-2206: E2E, Fixed problems with number inputs and file uploads.

# [AU-2205](https://helsinkisolutionoffice.atlassian.net/browse/AU-2205), [AU-2206](https://helsinkisolutionoffice.atlassian.net/browse/AU-2206)

## What was done
We've had a long lasting problem where data entered to various input fields would be missing part of its value when validation was being performed. I did some digging and came to the conclusion that this most likely has to do with page [Hydration](https://playwright.dev/docs/navigations#hydration). My guess is that Playwright would start filling in values to the form before the form and its elements were fully loaded, leading to a state of flaky inputs. I have added various `.waitFor({state: 'visible'})` implementations to the listed functions below in order to ensure that the elements are fully loaded before data is being entered.

## What has changed

- Refactored `clickButton()` (waiting for visible state).

- Refactored `data-drupal-selector-sequential` (waiting for visible state).

- Refactored `fillDynamicMultiValueField` and `fillMultiValueField`  (waiting for visible state).

- Added various `page.waitForLoadState()` implementations before filling out a form page.

- Refactored `hideSlidePopup` and removed unnecessary calls to it.

- Removed the default value for `filePath` in `uploadFile()` (seems to have caused a bug where the wrong file would sometimes be uploaded).

- Fixed a few tests that had incorrect data or were entering data with the wrong methods.

## How to install

Make sure your instance is up and running on correct branch.
* `git checkout feature/AU-2205-e2e-entered-numbers-bug`
* `make fresh`
* `make drush-cr`

## How to test
- Add this to the `.test_env` file so that only draft tests run (we only care about draft tests since those are the ones that get validated): 
`DISABLED_FORM_VARIANTS="success", "missing_values", "wrong_values", "under5000", "wrong_email", "wrong_email_2", "wrong_email_3"`

- Run `make test-pw` and let all the draft tests run. Make sure all pass.

![Screenshot 2024-03-04 at 17 45 13](https://github.com/City-of-Helsinki/hel-fi-drupal-grants/assets/117807015/7df85417-7da2-4537-adb0-7f30d62006aa)


[AU-2205]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2205?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AU-2206]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2206?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
